### PR TITLE
Support for nodes that crash

### PR DESCRIPTION
**Pull request by mystenmark**

### CODE REVIEW

File: msim/src/sim/net/mod.rs [hunk 1]
```diff
@@ -905,6 +905,7 @@ impl NetSim {
         debug!("delete_node {id}");
         let mut network = self.network.lock().unwrap();
         network.delete_node(id);
+        drop(network);
```
Comment (williampsmith) [inline msim/src/sim/net/mod.rs:908]: Why this line?

File: msim/src/sim/task.rs [hunk 2]
```diff
@@ -65,6 +67,49 @@ pub(crate) struct NodeInfo {
...
+}
+
+/// Kill the current node by panicking with a special type that tells the executor to kill the
+/// current node instead of terminating the test.
+pub fn kill_current_node(restart_after: Option<Duration>) {
+    let handle = runtime::Handle::current();
+    let cur_node_id = context::current_node();
+
+    let restart_after = restart_after.unwrap_or_else(|| {
+        Duration::from_millis(handle.rand.with(|rng| rng.gen_range(1000..3000)))
```
Comment (williampsmith) [inline msim/src/sim/task.rs:101]: Is the random time just to make the network seem more organic in its recovery?

**Review by williampsmith — APPROVED** [review]

File: msim/src/sim/net/mod.rs (hunk 1 above)
Comment (mystenmark) [reply msim/src/sim/net/mod.rs:908]: must release the lock before deleting the node from host_state below - that can call drop handlers which try to acquire this lock, which causes deadlocks.

File: msim/src/sim/task.rs (hunk 2 above)
Comment (mystenmark) [reply msim/src/sim/task.rs:101]: pretty much, yes.

**Pull request merged**